### DISCUSSION TITLE
Efficient implementation of SSIM

### DISCUSSION THREAD

**Discussion by ragavvenkatesan** [Lightning-AI/torchmetrics#973] [CompVision] [answered]
Is there any particular reason for storing the `preds` and `target` in SSIM and MSSSIM in [here](https://github.com/PyTorchLightning/metrics/blob/10b3a8529311579b3db1330842f730a643533065/torchmetrics/image/ssim.py#L116), instead of storing computed and total? We could just compute the averages in `compute` method and calculate the actual SSIM and MSSSIM on a mini-batch basis on the `update` method and keep track. The reason why we might want this is because in its current form, SSIM and MSSSIM cannot be tracked during say validation or in any cases, where we do not want to reset them at in the `_step_end` methods and instead reset them in the `_epoch_end` methods.

Thanks, R

**Comment by SkafteNicki** [comment]
Hi @ragavvenkatesan,
I do not think that there is any particular reason that `preds` and `target` is stored. Would you be willing sending a PR fixing this?

**Reply by ragavvenkatesan** [reply]
I could. Sorry I was out for a while. I can try and help fixing this.

**Comment by Borda** [comment, marked as answer]
@SkafteNicki acknowledged the performance concern. As of v1.9.0, here are the practical options for faster SSIM:

**1. Use a uniform kernel instead of Gaussian** (saves the Gaussian convolution):
```python
from torchmetrics.image import StructuralSimilarityIndexMeasure

ssim = StructuralSimilarityIndexMeasure(
    gaussian_kernel=False,
    kernel_size=11,
    data_range=1.0,
)
```

**2. Reduce kernel size** (default is 11):
```python
ssim = StructuralSimilarityIndexMeasure(kernel_size=7, data_range=1.0)
```
Smaller kernel = fewer convolution ops. For most use cases `kernel_size=7` is sufficient.

**3. Downscale before computing** (for high-res images):
```python
import torch.nn.functional as F

# Downsample 2x before SSIM — often negligible quality difference
preds_down = F.interpolate(preds, scale_factor=0.5, mode="bilinear")
target_down = F.interpolate(target, scale_factor=0.5, mode="bilinear")
score = ssim(preds_down, target_down)
```

**4. Use the functional interface** for one-off computations (avoids metric state overhead):
```python
from torchmetrics.functional.image import structural_similarity_index_measure

score = structural_similarity_index_measure(preds, target, data_range=1.0)
```

**5. Consider MS-SSIM** if you're already downscaling — it computes SSIM at multiple scales natively and can be more robust:
```python
from torchmetrics.image import MultiScaleStructuralSimilarityIndexMeasure

ms_ssim = MultiScaleStructuralSimilarityIndexMeasure(data_range=1.0)
```

The implementation itself has been optimized over the years (depthwise separable convolutions, etc.), so if you last benchmarked in 2022, it's worth retesting with the latest version.

Docs: [SSIM](https://lightning.ai/docs/torchmetrics/stable/image/structural_similarity.html) | [MS-SSIM](https://lightning.ai/docs/torchmetrics/stable/image/multi_scale_structural_similarity.html)
